src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Runtime.InteropServices;
+
+using CFStringRef = System.IntPtr;
+using CFArrayRef = System.IntPtr;
+using FSEventStreamRef = System.IntPtr;
+using size_t = System.IntPtr;
+using FSEventStreamEventId = System.UInt64;
+using CFTimeInterval = System.Double;
+using CFRunLoopRef = System.IntPtr;
+
+namespace System.IO
+{
+    // OSX implementation of FileSystemWatcher's PAL.  The main partial definition of FileSystemWatcher expects:
+    // - StartRaisingEvents
+    // - StopRaisingEvents
+    // - FinalizeDispose
+
+    public partial class FileSystemWatcher
+    {
+        // OS X has a case-sensitive File System
+        internal const bool CaseSensitive = true;
+
+        // Flags used to create the event stream
+        private const Interop.CoreServices.FSEventStreamCreateFlags EventStreamFlags = (Interop.CoreServices.FSEventStreamCreateFlags.kFSEventStreamCreateFlagFileEvents |
+                                                                                        Interop.CoreServices.FSEventStreamCreateFlags.kFSEventStreamCreateFlagWatchRoot);
+
+        // The last directory we were told to watch. This allows us to be smart about how we create streams
+        // and to only teardown and recreate streams when we need to.
+        private String _lastWatchedDirectory = String.Empty;
+
+        // The bitmask of events that we want to send to the user
+        private Interop.CoreServices.FSEventStreamEventFlags _filterFlags;
+
+        // The EventStream to listen for events on
+        private FSEventStreamRef _eventStream = IntPtr.Zero;
+
+        // The background thread to use to monitor events
+        private Thread _watcherThread = null;
+
+        // A reference to the RunLoop that we can use to start or stop a Watcher
+        private CFRunLoopRef _watcherRunLoop = IntPtr.Zero;
+
+        // Use an event to try to prevent StartRaisingEvents from returning before the
+        // RunLoop actually begins. This will mitigate a race condition where the watcher
+        // thread hasn't completed initialization and stop is called before the RunLoop even starts.
+        private AutoResetEvent _runLoopStartedEvent = new AutoResetEvent(false);
+
+        /// <summary>Called when FileSystemWatcher is finalized.</summary>
+        private void FinalizeDispose()
+        {
+            // Make sure we cleanup
+            StopRaisingEvents();
+            CleanupStreamAndWatcher();
+        }
+
+        private void StartRaisingEvents()
+        {
+            // Make sure the Start and Stop can be called from different threads and don't 
+            // stomp on the other's operation. We use the _lastWatchedDirectory instead of the
+            // RunLoop or StreamRef because IntPtrs are value types and can't be locked
+            lock (_lastWatchedDirectory)
```

Comment:
Locking on strings is generally considered a no-no (because of things like string interning, shared string constants, and optimizations based on strings being immutable, you can end up sharing a lock with some other completely unrelated piece of code), further so because in this case _lastWatchedDirectory can change.